builtin_array.go

```diff
@@ -1023,6 +1023,60 @@ func (r *Runtime) arrayproto_findIndex(call FunctionCall) Value {
 	return intToValue(-1)
 }
 
+func (r *Runtime) arrayproto_flat(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	l := toLength(o.self.getStr("length", nil))
+	depthNum := int64(1)
+	if len(call.Arguments) > 0 {
+		depthNum = call.Argument(0).ToInteger()
+	}
+	a := arraySpeciesCreate(o, 0)
+	r.flattenIntoArray(a, o, l, 0, depthNum, nil, nil)
+	return a
+}
+
+func (r *Runtime) flattenIntoArray(target, source *Object, sourceLen, start, depth int64, mapperFunction func(FunctionCall) Value, thisArg Value) int64 {
+	targetIndex, sourceIndex := start, int64(0)
+	for sourceIndex < sourceLen {
+		p := intToValue(sourceIndex)
+		if source.hasProperty(p) {
+			element := source.get(p, source)
+			if mapperFunction != nil {
+				element = mapperFunction(FunctionCall{
+					This:      thisArg,
+					Arguments: []Value{element, p, source},
+				})
+			}
+			if depth > 0 && isArray(element.ToObject(r)) {
+				elementLen := toLength(element.ToObject(r).self.getStr("length", nil))
+				targetIndex = r.flattenIntoArray(target, element.ToObject(r), elementLen, targetIndex, depth-1, mapperFunction, thisArg)
+			} else {
+				if targetIndex > maxInt {
+					r.typeErrorResult(true, "Invalid array length")
+					panic("unreachable")
```

Comment:
```suggestion
					panic(r.NewTypeError("Invalid array length"))
```

Reply:
I am fine with the change but just wanted to point out I copied it from 
https://github.com/dop251/goja/blob/51b43d63272ebd474a52ab0069fc510a8e1715e1/builtin_array.go#L114-L115

Reply:
I know 😞 Thanks.